linkerd/channel/Cargo.toml

```diff
@@ -0,0 +1,13 @@
+[package]
+name = "linkerd2-channel"
+version = "0.1.0"
+authors = ["Linkerd Developers <cncf-linkerd-dev@lists.cncf.io>"]
+edition = "2018"
+publish = false
+description = """
+A bounded MPSC channel where senders expose a `poll_ready` method.
```

Comment:
Why did tokio drop support for poll_ready? Do we understand the thought process behind that?

If we're going to implement our own, we should probably get some tests on it...

Also, would prefer that we do this in its own change (i.e. first do this against main with tokio-0.2, and then port to tokio-0.3 in this branch).

Reply:
> Why did tokio drop support for poll_ready? Do we understand the thought process behind that?

@olix0r I'm happy to provide some insight into this. Most of the Tokio synchronization primitives lost their `poll`-based interfaces in 0.3 as part of the move to intrusive lists of wakers for synchronization primitives (see tokio-rs/tokio#2325, tokio-rs/tokio#2509, and tokio-rs/tokio#2861). This change takes advantage of the inherently pinned nature of `async fn` and `async` blocks to avoid needing a separate heap allocation to store the waiter state for a task waiting on a synchronization primitive. However, it means that a synchronization primitive can _only_ be waited on when the future that waits on it is pinned --- otherwise, there is a potential dangling pointer. The `poll`-based APIs allowed waiting on synchronization primitives from unpinned contexts, so they were removed. 

To wait on the synchronization primitives from contexts that may not be pinned, such as `poll_ready`, it's necessary to add a `Pin<Box<...>>` around the future that's waiting on the synchronization primitive. This ensures that the future will not move while it's part of the wait list. It's important to note that this isn't an _additional_ allocation per waiter versus Tokio 0.2; instead, it's the same allocation that would have _always_ happened internally to the synchronization primitive in the 0.2 API. Now, it's moved outside of the `tokio::sync` type so that it can be avoided when used with `async`/`await` syntax, and added by the user when polling the sync primitives.

Reply:
> If we're going to implement our own, we should probably get some tests on it...
> 
> Also, would prefer that we do this in its own change (i.e. first do this against main with tokio-0.2, and then port to tokio-0.3 in this branch).

I'm happy to try and factor this out into a separate PR, and add tests. It's worth noting that I used approximately the same approach in tower-rs/tower#476, for `tower-buffer`; it's necessary to reimplement it in the proxy codebase because we use our own buffer implementation.

Also, it's possible that there will eventually be a wrapper for MPSC senders that adds back `poll_ready`, in `tokio-util`: https://github.com/tokio-rs/tokio/pull/3105. This is moved out of the core MPSC API so that the additional overhead is opt-in. However, I'm not sure when that PR will be ready to merge...